server/src/main/java/org/elasticsearch/repositories/RepositoriesService.java

```diff
@@ -349,9 +353,76 @@ public Repository repository(String repositoryName) {
         if (repository != null) {
             return repository;
         }
+        repository = internalRepositories.get(repositoryName);
+        if (repository != null) {
+            return repository;
+        }
         throw new RepositoryMissingException(repositoryName);
     }
 
+    public void registerInternalRepository(String name, String type, Settings settings) {
+        RepositoryMetaData metaData = new RepositoryMetaData(name, type, settings);
+        Repository existingRepository = internalRepositories.get(name);
+        if (existingRepository != null && existingRepository.getMetadata().equals(metaData)) {
+            // Identical repository already exists, no need to update it.
+            return;
+        }
+
+        // TODO: Normally we would do validation when we update a repository to ensure that it is not in use.
+        //  Are we okay with not including that validation under the assumption that internal operations
+        //  will do the right thing.
```

Comment:
I think it's fine not having this validation here. If we remove the settings parameter from this method, a repo will never be updated with different settings.